Step2 Spring MVC - 사용자 인증

JoisFe opened the pull request.

JoisFe committed:
ste1 reservation repository cherrypick

JoisFe committed:
STEP1 방탈출 예약 관리 피드백 반영
- dto -> domain, domain -> dto 매핑을 adator -> service layer에서 하도록 수정
- 비즈니스 로직과 관련된 검증을 service layer 에서 하도록 수정

JoisFe committed:
STEP1 방탈출 예약 페이지 노출

JoisFe committed:
step2 - 1단계 로그인 구현
feat: 사용자 예약 페이지 연동
- 사용자 예약 페이지 조회 API
- 예약 가능 시간 조회 API 구현

JoisFe committed:
step2 - 1단계 로그인 구현
feat : 토큰을 이용한 로그인 구현
- 로그인 요청시 로그인 정보를 통해 토큰 생성 -> 쿠키에 저장
- 쿠키 값을 통해 로그인 정보(인증 정보) 를 조회

isfejo committed:
step2 Spring MVC - 사용장 인증 : 로그인 리팩터링

- 예약 새성 기능 변경
- HandlerMethodArgumentResolver 활용한 쿠키 통한 멤버 조회

isfejo committed:
step2 Spring MVC - 사용장 인증 : 로그인 리팩터링

- interceptor를 활용한 ADMIN role 권한을 가진 member만 /admin/** 에 접근 가능하도록 변경
- member 에 role 추가 및 jwt claim에 role 추가

Hyeon9mak commented on src/main/java/roomescape/adapter/in/web/LoginController.java:
```diff
@@ -0,0 +1,52 @@
+import roomescape.application.dto.MemberCommand;
+import roomescape.application.dto.MemberResponse;
+import roomescape.application.port.in.LoginUseCase;
+
+@Controller
+@RequestMapping("/login")
+public class LoginController {
+
+  private final LoginUseCase loginUseCase;
+
+  public LoginController(LoginUseCase loginUseCase) {
+    this.loginUseCase = loginUseCase;
+  }
+
+  @GetMapping
+  public String getLoginPage() {
+    return "login";
+  }
+
+  @ResponseBody
+  @GetMapping("/check")
+  public ResponseEntity<MemberResponse> checkInvalidLogin(@UserInfo MemberResponse memberResponse) {
+    return ResponseEntity.ok(memberResponse);
+  }
```
`@Controller` 와 `@RestController` 의 차이는 무엇일까요?!
(이미 잘 알고 사용하신듯 하지만 ㅎㅎ)

JoisFe replied:
<img width="458" alt="image" src="https://github.com/next-step/spring-roomescape-auth/assets/90208100/3b6e7337-52c4-493d-b363-8bf1d52e656d">

`@RestController 경우` `@Controller`에 `@ResponseBody` 만 추가된 어노테이션입니다.

`@ResponseBody` 경우
<img width="659" alt="image" src="https://github.com/next-step/spring-roomescape-auth/assets/90208100/1d73ad21-2936-40f9-bc4f-71e5a04e9f47">

https://docs.spring.io/spring-framework/reference/web/webflux/controller/ann-methods/responsebody.html

HttpMessageWriter를 통해 응답 본문에 반환을 직렬화하여 반환한다고 합니다.

<img width="850" alt="image" src="https://github.com/next-step/spring-roomescape-auth/assets/90208100/39e1ab4d-35ea-449d-9a0c-c4e5b5ff0037">

https://www.baeldung.com/spring-request-response-body

Hyeon9mak commented on src/main/java/roomescape/adapter/in/web/LoginController.java:
```diff
@@ -0,0 +1,52 @@
+  @ResponseStatus(HttpStatus.OK)
+  @ResponseBody
+  @PostMapping("")
+  public void login(@RequestBody LoginCommand loginCommand, HttpServletResponse response) {
+    String jwt = loginUseCase.createToken(loginCommand);
+
+    Cookie jwtCookie = new Cookie("jwt", jwt);
+    jwtCookie.setHttpOnly(true);
+    jwtCookie.setPath("/");
+    response.addCookie(jwtCookie);
+  }
```
쿠키를 통한 인증이 무한히 진행될 수도 있겠네요!
쿠키가 유효할 수 있는 기간을 별도로 설정해주는 건 어떨까요?

JoisFe replied:
넵 쿠기 유효 기간을 설정하도록 하겠습니다.
감사합니다.

Hyeon9mak commented on src/main/java/roomescape/adapter/in/web/advice/ExceptionController.java:
```diff
@@ -0,0 +1,45 @@
+//import org.springframework.web.bind.annotation.ExceptionHandler;
+//import org.springframework.web.bind.annotation.RestControllerAdvice;
+//import roomescape.exception.AuthenticationException;
+//import roomescape.exception.AuthorizationException;
```
주석처리 해두신 이유가 있을까요?!

JoisFe replied:
앗 이 부분은 제가 주석을 다시 풀었어야했는데 그대로 제출한 것 같습니다. 주석풀도록 하겠습니다.

Hyeon9mak commented on src/main/java/roomescape/adapter/out/MemberEntity.java:
```diff
@@ -0,0 +1,38 @@
+package roomescape.adapter.out;
+
+public class MemberEntity {
+
+    private final Long id;
+    private final String name;
+    private final String email;
+    private final String password;
+    private final String role;
+
+    public MemberEntity(Long id, String name, String email, String password, String role) {
+        this.id = id;
+        this.name = name;
+        this.email = email;
+        this.password = password;
+        this.role = role;
+    }
+
+    public Long getId() {
+        return id;
+    }
+
```
id 를 통해 구분이 되는 entity(= identifier + vo) 라면, id 에만 동등성을 부여해 다뤄볼 수도 있을거 같아요!

JoisFe replied:
넵 id 가 unique하기에 id 에 따라 동등성을 부여하도록 적용하겠습니다.

Hyeon9mak commented on src/main/java/roomescape/adapter/out/ReservationEntity.java:
```diff
@@ -0,0 +1,42 @@
+    return theme;
+  }
+
+  public static ReservationEntity of(Long id, String name, String date, ReservationTimeEntity time) {
+    return new ReservationEntity(id, name, date, time, null);
+  }
```
mapper 클래스가 존재하는 것 같았는데! 별도의 팩터리 메서드를 만들어주신 이유가 있을까요?

Hyeon9mak commented on src/main/java/roomescape/adapter/out/in_memory/ReservationInMemoryRepository.java:
```diff
@@ -0,0 +1,69 @@
+import roomescape.domain.ReservationTime;
+import roomescape.domain.Theme;
+
+@Component
+public class ReservationInMemoryRepository implements ReservationPort {
+
+  private final AtomicLong index;
+
+  private final Map<Long, ReservationEntity> reservationMap;
+
+  public ReservationInMemoryRepository() {
+    index = new AtomicLong(0);
+    reservationMap = new HashMap<>();
+
+    ReservationTime reservationTime1 = new ReservationTime(1L, "12:30");
+    ReservationTime reservationTime2 = new ReservationTime(2L, "11:00");
+    saveReservation(new Reservation(null, "브라운", "2024-06-06", reservationTime1, null));
+    saveReservation(new Reservation(null, "블랙", "2024-07-06", reservationTime2, null));
+  }
```
테스트를 위해 작성된 클래스로 보이네요!
main 대신 test 패키지에 위치시키는 건 어떨까요?

JoisFe replied:
앗 해당 부분이 기존에 JDBC Template 적용 이전에
Java 자료구조를 이용해서 데이터 저장 및 조회를 작업한 내용입니다.

Hyeon9mak replied:
엇! 더 이상 사용되지 않는 코드라면 제 코멘트와 같이 오해를 할 수도 있으니 시원하게 제거하는 건 어떨까요?! 😄

Hyeon9mak commented on src/main/java/roomescape/adapter/out/jdbc/MemberJdbcRepository.java:
```diff
@@ -0,0 +1,77 @@
+        this.jdbcTemplate = jdbcTemplate;
+    }
+
+    @Override
+    public List<Member> findMembers() {
+        String sql = "SELECT id, name, email, role FROM member";
+
+        List<MemberEntity> memberEntities = jdbcTemplate.query(
+            sql, (resultSet, rowNum) -> {
+                MemberEntity memberEntity = new MemberEntity(
+                    resultSet.getLong("id"),
+                    resultSet.getString("name"),
+                    resultSet.getString("email"),
+                    null,
+                    resultSet.getString("role")
+                );
+
+                return memberEntity;
+            }
+        );
+
+        return memberEntities.stream()
+                             .map(MemberMapper::mapToDomain)
+                             .toList();
+    }
```
계속해서 MemberEntity 를 만들기 위한 코드가 반복되고 있네요!
row mapper 를 repository 의 멤버 변수로 분리해보는 건 어떨까요?

JoisFe replied:
findMembers와 findMemberByEmail  모두 MemberEntity 를 만들고자하는데요
findMembers는 findMemberByEmail과 다르게 password를 매핑하고 싶지 않아
row mapper 두가지 상황을 모두 만들기 위해선 변수 2개가 생겨야 되는 것 같아 그냥 람다로 상황에 맞게끔 구현하였습니다.

매핑하는 상황이 다를때도 1가지 변수로 처리 가능한지 여쭙고 싶습니다.

Hyeon9mak replied:
오 아닙니다! 제가 꼼꼼히 읽지 않고 코멘트를 남겼네요 🙏

Hyeon9mak commented on src/main/java/roomescape/adapter/out/jdbc/MemberJdbcRepository.java:
```diff
@@ -0,0 +1,77 @@
+    public Optional<Member> findMemberByEmail(String email) {
+        String sql = "SELECT id, name, email, password, role FROM member WHERE email = ?";
+
+        List<MemberEntity> memberEntities = jdbcTemplate.query(
+            sql, new Object[] {email}, (resultSet, rowNum) -> {
+                MemberEntity member = new MemberEntity(
+                    resultSet.getLong("id"),
+                    resultSet.getString("name"),
+                    resultSet.getString("email"),
+                    resultSet.getString("password"),
+                    resultSet.getString("role")
+                );
+
+                return member;
+            }
+        );
+
```
단건 조회에 대해서는 queryObject 로 조회하는 건 어떨까요?

JoisFe replied:
queryObject로 단일 조건시 1개도 나오지 않는 경우 에러가발생하더라구요, 
저는 해당 메서드를 호출한 serviceLayer에 1건이 나오지 않더라도 Optional.ofNullable 로 감싸서 반환 후
해당 layer에서 Optional내 값을 get할때 orElseThrow로 원하는 에러를 던지고 싶어 
해당 query 메서드를 사용했습니다.

queryObject 결과시 IncorrectResultSizeDataAccessException 가 발생하던데, 혹시 해당 에러를 catch하여 
Optional.ofNullable로 사용하는 방법이 더 괜찮은 방법일까요?

Hyeon9mak replied:
말씀해주신 방법도 충분히 좋다고 생각됩니다!
다만 1개를 초과하는 데이터가 조회되는 경우에 대해서도 관리를 해주어야 할 것 같은데요,
이런 경우 isEmpty 과 더불어 count(혹은 length)가 1인지 검증을 해주어야겠네요!

Hyeon9mak commented on src/main/java/roomescape/adapter/out/jdbc/ReservationJdbcRepository.java:
```diff
@@ -0,0 +1,126 @@
+    return Optional.ofNullable(reservation);
+  }
+
+  @Override
+  public Reservation saveReservation(Reservation reservation) {
+
+    String sql = "INSERT INTO reservation(name, date, time_id, theme_id) VALUES(?, ?, ?, ?)";
+    KeyHolder keyHolder = new GeneratedKeyHolder();
+
+    jdbcTemplate.update(connection -> {
+      PreparedStatement ps = connection.prepareStatement(sql, Statement.RETURN_GENERATED_KEYS);
+      ps.setString(1, reservation.getName());
+      ps.setString(2, reservation.getDate());
+      ps.setLong(3, reservation.getTime().getId());
+      ps.setLong(4, reservation.getTheme().getId());
+      return ps;
+    }, keyHolder);
+
+    return reservation.addId(Objects.requireNonNull(keyHolder.getKey())
+                                    .longValue());
+  }
```
KeyHolder 를 활용해주셨군요! Member 에 대해서도 비슷한 동작을 보장해주면 좋겠네요 👍

Hyeon9mak replied:
SimpleJdbcInsert 를 활용해보아도 좋을거 같아요 😄

JoisFe replied:
앗 멤버에도 적용하도록 하겠습니다.

그리고 insert시에 생성되는 id 값을 가져오기 위해 찾아보다가 KeyHolder를 이용하게되었는데, SimpleJdbcInsert를 리뷰를 통해 알게되었습니다.
사용법을 확인해보니 훨씬 간단한 것 같네요. 적용해보도록 하겠습니다.

Hyeon9mak commented on src/main/java/roomescape/adapter/out/jdbc/ThemeJdbcRepository.java:
```diff
@@ -0,0 +1,105 @@
+package roomescape.adapter.out.jdbc;
+
+import java.sql.PreparedStatement;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import org.springframework.jdbc.core.JdbcTemplate;
+import org.springframework.jdbc.support.GeneratedKeyHolder;
+import org.springframework.jdbc.support.KeyHolder;
+import org.springframework.stereotype.Repository;
+import roomescape.adapter.mapper.ThemeMapper;
+import roomescape.adapter.out.ThemeEntity;
+import roomescape.application.port.out.ThemePort;
```
하나의 모듈 내에서 패키지를 이와 같이 분리해주신 이유가 있을까요?
interface 에 대해 port 라고 명칭을 지어주신 후, 구현체의 명칭을 repository 로 정하신 이유가 궁금해요!

<img width="726" alt="image" src="https://github.com/next-step/spring-roomescape-auth/assets/37354145/391ad1ee-df4f-4bf9-abaa-bdb0daad5732">

`@Repository` 어노테이션의 설명을 살펴보면 위와 같은데, 
인터페이스의 명칭을 repository 로 사용하고, 구현체의 이름을 다른 걸로 바꾸는게 더 자연스러울 수도 있겠네요 😄

JoisFe replied:
애플리케이션 계층에 아웃고잉 포트 (port.out)를 통해 외부에서 (DB 쿼리 혹은 다른 애플리케이션 서버에 API 요청) 원하는 것을 가져오기 위한 인터페이스를 명세하고 
어댑터 계층에 아웃고잉 어댑터 (adapter.out)에 해당 구현체를 구현하였습니다.

인터페이스에 대한 구현체가 여러가지가 될 수 있다고 판단하여 인터페이스명을 넓은 의미로 ~~Port 라고 명시를 했습니다.

다만 말씀주신대로 포트말고 어떤 목적을 가지는 인터페이스인지 전혀 알 수 없네요..

궁금한점이 원하는 정보를 Repository (storage) 에서 가져오지 않고 다른 애플리케이션 서버에 API를 통해 가져오는 구현이 생길수도 있다고 생각이 드는데요, 그러한 경우도 Repository로 인터페이스 명을 사용해도 되는지 여쭙고 싶습니다.

Hyeon9mak replied:
> 원하는 정보를 Repository (storage) 에서 가져오지 않고 다른 애플리케이션 서버에 API를 통해 가져오는 구현이 생길수도 있다고 생각이 드는데요, 그러한 경우도 Repository로 인터페이스 명을 사용해도 되는지 여쭙고 싶습니다

repository 는 "내가 원하는 정보를 어디서든 일단 가져만 와봐. 내 관심사는 그걸로 무얼 할것인가야." 가 녹아든 네이밍인 것 같아요.
즉 어딘가 저장소라는 것에만 집중할 뿐, 그 대상이 db 인지, 외부 서버인지, excel 인지, 인메모리 공간인지는 관심사가 아니라고 생각됩니다.

그래서 repository 라는 common 한 명칭을 사용하는 것 같아요!

JoisFe replied:
넵 이해했습니다. 
감사합니다.
다음 step시작시에 반영하도록 하겠습니다.

Hyeon9mak commented on src/main/java/roomescape/application/port/in/LoginUseCase.java:
```diff
@@ -0,0 +1,16 @@
+
+  MemberResponse findMember(String payload, Role role);
+
+  MemberResponse findMemberByJwt(String jwt);
+}
```
유저를 식별하는 정보가 jwt 토큰이라는 건 세부 구현사항 아닐까요?
jwt 라는 명칭 대신 더 제네럴한 명칭을 사용하는 건 어떨까요? 😉

JoisFe replied:
제가 바로 위 댓글에 답변이 단 생각과 완전히 상충되게 메서드 명을 정했네요..
인터페이스에서 너무 특정 구현에 맞는 메서드 명인 것 같습니다. 적합한 추상적인 메서드명으로 교체해 보도록 하겠습니다.

Hyeon9mak commented on src/main/java/roomescape/application/port/in/MemberUseCase.java:
```diff
@@ -0,0 +1,12 @@
+package roomescape.application.port.in;
+
+import java.util.List;
+import roomescape.application.dto.MemberCommand;
+import roomescape.application.dto.MemberResponse;
+
+public interface MemberUseCase {
+
+    List<MemberResponse> findMembers();
+
+    void registerMember(MemberCommand memberCommand);
+}
```
아마도 헥사고날 아키텍처를 적용하고자 하셨던걸로 이해되는데, 헥사고날 아키텍처에서 유스케이스를 표현하는 인터페이스는 말 그대로 하나의 유스케이스의 동작만 갖고 있어야 한다고 생각해요!

현재는 멤버정보 리스트를 조회할 때 1개, 멤버를 새로 가입시킬 때 2개의 동작을 갖고 있는 걸로 보이네요...
재철님은 어떻게 생각하실까요?

JoisFe replied:
해당 부분이 고민이 많았습니다.
헥사고날 아키텍처 예시를 보면 하나의 동작에 따라 유스케이스를 만들고 구현체를 만들더라구요
그런데 이렇게하다보니 너무많은 유스케이스가 생기는 것 같아 도메인 기준으로 유스케이스 하나에 많은 동작을 넣었습니다.

굳이 동작별로 유스케이스 인터페이스를 나눠야할까? 라는 생각에 알아보니 (만들면서 배우는 클린 아키텍처 - 톰 홈버그)
현재 하나의 유스케이스에 여러 동작을 담고 있으면
layered architecture 문제점인
1. 애플리케이션의 기능 조각 혹은 특성을 구분 짓는 패키지 경계가 없는 점
2. 애플리케이션이 어떤 유스케이스들을 제공하는지 파악할 수 없는점

2에 해당 하는 것 같습니다.

Layered Architecture에서 단순히 AccountService -> 어떤 유스케이스를 구현했는지 파악이 힘들고 해당 인터페이스를 어떤 구현체가 어떤 것을 구현했는지 추측하거나 힘들게 찾아야 되는 상황이 생긴다고 합니다. (표현력이 떨어지는 구조)
또한 유스케이스 라는 용어를 제대로 이해하지 못하고 헥사고날 아키텍처를 사용한 것 같습니다.

저는 현재 각 도메인마다 많은 기능이 없어서 해당 문제점에 크게 느끼지는 못하고 있었지만 각 도메인 마다 많은 기능이 생긴다면 매우 지금 방식대로 진행한다면 추후에 수정해야할 기능을 찾는데 너무 많은 시간을 소비하게 될 것 같습니다.

Hyeon9mak replied:
👍 👍 어떤 부분이 문제가 될지 예상까지 해주셨군요! 💯 
답변 주신 내용만으로도 충분하다고 생각됩니다 크크 👍 👍 

만약~ 현재 유스케이스가 "회원의 정보를 조회해주세요. 회원이 아니라면 가입시켜주세요." 와 같았다면 `findMembers`, `registerMember` 가 하나의 인터페이스에 명세되어 있는 것이 자연스러웠겠네요!

요런 여러가지 고민을 해보시는 것도 재밌을거 같습니다 ㅎㅎ

Hyeon9mak commented on src/main/java/roomescape/application/service/ReservationService.java:
```diff
@@ -0,0 +1,93 @@
+        this.themePort = themePort;
+    }
+
+    @Transactional(readOnly = true)
+    @Override
+    public List<ReservationResponse> retrieveReservations() {
+        return reservationPort.findReservations()
+                              .stream()
+                              .map(ReservationMapper::mapToResponse)
+                              .toList();
+    }
```
`@Transactional` 내부에서 `@Transactional(readOnly = true)` 를 명세하는 것과
`@Transactional(readOnly = true)` 만 명세하는 것에는 어떤 차이가 있을까요?

JoisFe replied:
특정 메서드에 `@Transactinoal(readOnly = true)` 만을 명세하는 경우 해당 클래스에 다른 메서드에는 트랜잭션 관리를 하지 않게 됩니다.

메서드에 트랜잭션 관리가 필요하거나 읽기 전용 트랜잭션을 적용을 하나하나 적용하는 것보다
클래스에 `@Transcation`을 달아 전체 메서드에 트랜잭션 적용을 해두고 읽기 전용 트랜잭션이 필요한 메서드에 `@Transaction(readOnly =true)`를 달아 오버라이드 하도록 하였습니다.

Hyeon9mak reviewed (changes requested):
안녕하세요 재철님! spring mvc 미션을 함께할 최현구입니다. 🙇‍♂️ 
사용자 인증 미션 잘 진행해주셨습니다!

패키지 구조를 매우 잘게 분리해주셨네요!
관련하여 고민하실 만한 포인트들 코멘트 남겼습니다!
궁금하신 점 편하게 질문 남겨주세요!!

JoisFe committed:
step2 Spring MVC - 사용자 인증: 로그인 리팩터링

- cookie max age 설정 추가
- entity 객체 id값으로 동등성 부여
- jdbc insert 경우 PreparedStatement, KeyHolder 방식 -> SimpleJdbcInsert 방식으로 변경
- findMemberByJwt -> 조금 더 추상적인 메서드 명으로 수정

JoisFe committed:
step2 Spring MVC - 사용자 인증 : 로그인 리팩터링
- 사용자 예약 -> timeId, themeId 를 통해 reservationTime, theme 을 조회하여 rservation 등록하도록 수정

JoisFe commented on src/main/resources/static/js/user-reservation.js:
```diff
@@ -151,8 +155,10 @@ function checkDateAndThemeAndTime() {
+      '.theme-slot.active')?.getAttribute('data-theme-id');
+  const selectedTimeId = document.querySelector(
+      '.time-slot.active')?.getAttribute('data-time-id');
   const name = document.getElementById('user-name').value;
```
id : user-name이 보이지 않는 것 같은데, name을 어떻게 받아올 수 있을까요?

Hyeon9mak replied:
로그인 후 회원정보 영역으로부터 끌어올 정보로 생각되는데... 회원가입 페이지가 준비되지 않아 정상적인 플로우 이용이 불가능하네요 😅 
이번 미션에서 중요한 것은 spring, server 관리 이므로 넘어가셔도 될거 같습니다!

Hyeon9mak commented on src/main/java/roomescape/adapter/in/web/LoginController.java:
```diff
@@ -0,0 +1,52 @@
+  public void login(@RequestBody LoginCommand loginCommand, HttpServletResponse response) {
+    String jwt = loginUseCase.createToken(loginCommand);
+
+    Cookie jwtCookie = new Cookie("jwt", jwt);
+    jwtCookie.setHttpOnly(true);
+    jwtCookie.setPath("/");
+    jwtCookie.setMaxAge(30 * 60);
+    response.addCookie(jwtCookie);
```
30*60 이 최종적으로 어느정도의 값인지 단위를 주석 등으로 표현해주어도 좋겠어요 😉

Hyeon9mak commented on src/main/java/roomescape/adapter/out/jdbc/ReservationJdbcRepository.java:
```diff
@@ -0,0 +1,127 @@
+    return Optional.ofNullable(reservation);
+  }
+
+  @Override
+  public Reservation saveReservation(Reservation reservation) {
+    SimpleJdbcInsert jdbcInsert = new SimpleJdbcInsert(jdbcTemplate)
+      .withTableName("reservation")
+      .usingGeneratedKeyColumns("id");
+
+    Map<String, Object> columnMap = new HashMap<>();
+    columnMap.put("name", reservation.getName());
+    columnMap.put("date", reservation.getDate());
+    columnMap.put("time_id", reservation.getTime().getId());
+    columnMap.put("theme_id", reservation.getTheme().getId());
+
+    Number reservationId = jdbcInsert.executeAndReturnKey(columnMap);
```
SimpleJdbcInsert 를 멤버변수로 갖도록 하는 것도 방법일거 같습니다 😉 👍

JoisFe replied:
1이 아닌경우가 생기지 않을거다라는 생각에 검증을 하지 않았지만 
확실히 validation은 해주는게 좋을 것 같네요! 다음 step 에 추가하도록 하겠습니다.

또한 SimpleJdbcInsert 또한 insert 마다 객체를 생성하기 보단 멤버변수로 가지고 있는게 좋아보이네요!! 감사합니다.

Hyeon9mak commented on src/main/java/roomescape/application/port/in/LoginUseCase.java:
```diff
@@ -0,0 +1,16 @@
+
+  MemberResponse findMember(String payload, Role role);
+
+  MemberResponse findMemberByToken(String jwt);
```
😉 👍 파라미터 명도 살짝 바뀌면 좋겠네요 크크~

JoisFe replied:
아이쿠 해당 부분도 반영 하겠습니다.

Hyeon9mak reviewed (approved):
안녕하세요 재철님! 피드백 잘 반영해주셨습니다.
다음 단계 진행에 참고하실만한 코멘트 남겼드렸습니다!
다음 단계에서 뵐게요 😄
